src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -230,14 +230,28 @@ private ServiceCallSite TryCreateExact(ServiceDescriptor descriptor, Type servic
             return null;
         }
 
-        private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, int slot)
+        private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, int slot, bool throwOnConstraintViolation)
         {
             if (serviceType.IsConstructedGenericType &&
                 serviceType.GetGenericTypeDefinition() == descriptor.ServiceType)
             {
                 Debug.Assert(descriptor.ImplementationType != null, "descriptor.ImplementationType != null");
                 var lifetime = new ResultCache(descriptor.Lifetime, serviceType, slot);
-                Type closedType = descriptor.ImplementationType.MakeGenericType(serviceType.GenericTypeArguments);
+                Type closedType;
+                try
+                {
+                    closedType = descriptor.ImplementationType.MakeGenericType(serviceType.GenericTypeArguments);
+                }
+                catch (ArgumentException ex)
+                {
+                    if (throwOnConstraintViolation)
```

Comment:
When does this happen? We aimed to have 0 first chance exceptions in ASP.NET Core. This introduces some....

Reply:
It does. There's unfortunately no way around this currently, but I have an API issue open to extend the `Type` API to have some kind of `Type.Can/TryMakeGenericType` method to use: https://github.com/dotnet/runtime/issues/28033 <- feel free to upvote :)

The behavior here we're trying to enforce is "if you ask for a single closed service, let the exception happen. If you ask for an enumerable, only return the ones that fit the constraint". The existing behavior throws *all* the time, this changes to let the exception get thrown *some* of the time.

Reply:
Will this make existing application registration behavior change at all? So I pay the cost of I haven't explicitly opted into this feature? Seems like yes but I want to confirm.

Reply:
This doesn't affect registration behavior, it's only on the resolution side. It's hard to compare performance of existing behavior since the existing behavior is an unhandled exception. On the previous PR I did do some benchmarking to compare the constrained vs. unconstrained paths:

https://gist.github.com/jbogard/a2be5b903c44c8b2b3c2ef0d1e8c1e0c

Reply:
I think we should re-throw the original exception. Today the container doesn't do think like make sure generic arity is the same. This will assume that every argument exception is a constraint mismatch when that isn't the case today.

We can also make this a bit more accurate by checking for those arity mismatches earlier but today we don't and this PR doesn't do that either. From the docs:

> The number of elements in typeArguments is not the same as the number of type parameters in the current generic type definition.

> -or-

> Any element of typeArguments does not satisfy the constraints specified for the corresponding type parameter of the current generic type.

>-or-

> typeArguments contains an element that is a pointer type (IsPointer returns true), a by-ref type (IsByRef returns true), or Void.

Reply:
Well it would simplify the approach just to re-throw instead of trying to assume. Not a problem.